src/services/item/plugins/html/h5p/index.ts

```diff
@@ -0,0 +1,225 @@
+  } = fastify;
+
+  validatePluginOptions(options);
+  const { routes, fileStorage } = options;
+  const { pathPrefix = '' } = fileStorage;
+
+  // todo: file service should be a singleton with mutliple storage drivers
+  const fileService = new FileService(fileStorage.config, fileStorage.type);
+
+  const h5pValidator = new H5PValidator();
+
+  const h5pService = new H5PService(fileService, pathPrefix, H5P_FILE_MIME_TYPE, h5pValidator);
+
+  // question: this is difficult to move this in the service because of the transaction
+  /**
+   * Creates a Graasp item for the uploaded H5P package
+   * @param filename Name of the original H5P file WITHOUT EXTENSION
+   * @param contentId Storage ID of the remote content
+   * @param remoteRootPath Root path on the remote storage
+   * @param member Actor member
+   * @param parentId Optional parent id of the newly created item
+   */
+  async function createH5PItem(
+    member: Member,
+    filename: string,
+    contentId: string,
+    parentId?: string,
+  ): Promise<Item> {
+    const metadata = {
+      name: h5pService.buildH5PPath('', filename),
+      type: ItemType.H5P,
+      extra: h5pService.buildH5PExtra(contentId, filename),
+    };
+    return db.transaction(async (manager) => {
+      return itemService.post(member, buildRepositories(manager), {
+        item: metadata,
+        parentId,
+      });
+    });
+  }
+
+  /**
+   * In local storage mode, proxy serve the H5P content files
+   * In the future, consider refactoring the fileService so that it can be grabbed from the
+   * core instance and can serve the files directly (with an option to use or not auth)
+   */
+  if (fileStorage.type === ItemType.LOCAL_FILE) {
+    /** Helper to set CORS headers policy */
+    const setHeaders = (response: FastifyStaticReply) => {
+      response.setHeader('Cross-Origin-Resource-Policy', 'same-site');
+    };
+
+    // serve integration html
+    const integrationRoute = path.join(
+      routes?.assets ?? DEFAULT_H5P_ASSETS_ROUTE,
+      'integration.html',
+    );
+    fastify.get(integrationRoute, async (req, res) => {
+      const html = renderHtml(
+        routes?.assets ?? DEFAULT_H5P_ASSETS_ROUTE,
+        routes?.content ?? DEFAULT_H5P_CONTENT_ROUTE,
+        hosts ?? ['localhost'],
+      );
+      res.send(html);
+    });
+
+    // hack to serve the "dist" folder of package "h5p-standalone"
+    const h5pAssetsRoot = path.dirname(require.resolve('h5p-standalone'));
+    fastify.register(fastifyStatic, {
+      root: h5pAssetsRoot,
+      prefix: routes?.assets ?? DEFAULT_H5P_ASSETS_ROUTE,
+      decorateReply: false,
+      setHeaders,
+    });
+
+    const h5pStorageRoot = path.join(fileStorage.config.local.storageRootPath, pathPrefix);
+    fs.mkdirSync(h5pStorageRoot, { recursive: true });
+    fastify.register(fastifyStatic, {
+      root: h5pStorageRoot,
+      prefix: routes?.content ?? DEFAULT_H5P_CONTENT_ROUTE,
+      decorateReply: false,
+      setHeaders,
+    });
+  }
```

Comment:
I guess the exact integration html and the folders to be served are H5P-specific. But in general, wouldn't all html services need a default way to serve the files statically w.r.t to the file service type (i.e. `local`), and ideally the child class should not have to perform the selection? One way to enforce that is to have an abstract function in the parent class `serve` which must be redefined in the children. Then the parent exposes a `handleServe` which performs the local check and calls the subclass's `serve`.

Optionally, it may also want to define a default way to render a templated page (e.g. the parent could "force" the child classes to use moustachejs). This can be done as a default implementation for a method e.g. `render` in the parent class.

Reply:
I hear your comment, but would you really put that code in the `service`?  It actually defines endpoints, maybe it makes more sense to create a sub-plugin `localHtmlPluginController`?

Reply:
True, i'm not sure about the exact boundaries here. But you could imagine that the html services should expose a `serve` method, which all html controllers should call (this can be enforced via an abstract parent html controller for instance)

Reply:
I think I need an example here^^'